File: band-limited-square-waves/GenerateSquareWave.java
/* 
 * Band-limited square waves (Java)
 * 
 * Copyright (c) 2016 Project Nayuki
 * All rights reserved. Contact Nayuki for licensing.
 * https://www.nayuki.io/page/band-limited-square-waves
 */

import java.io.BufferedOutputStream;
import java.io.DataOutput;
import java.io.DataOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;


public final class GenerateSquareWave {
	
	public static void main(String[] args) throws IOException {
		// Check number of command line arguments
		if (args.length != 5 && args.length != 6) {
			System.err.println("Usage: java GenerateSquareWave Frequency DutyCycle SampleRate Duration [BandLimited/Naive] Output.wav");
			System.err.println("Example: java GenerateSquareWave 440.0 0.5 48000 1.0 BandLimited Output.wav");
			System.exit(1);
			return;
		}
		
		// Get all command line arguments
		double frequency = Double.parseDouble(args[0]);
		double dutyCycle = Double.parseDouble(args[1]);
		int sampleRate = Integer.parseInt(args[2]);
		double duration = Double.parseDouble(args[3]);
		String mode;
		File outFile;
		if (args.length == 5) {
			mode = "BandLimited";
			outFile = new File(args[4]);
		} else if (args.length == 6) {
			mode = args[4];
			outFile = new File(args[5]);
		} else
			throw new AssertionError();
		
		// Check value ranges
		if (frequency <= 0)
			throw new IllegalArgumentException("Frequency must be positive");
		if (dutyCycle < 0 || dutyCycle > 1)
			throw new IllegalArgumentException("Duty cycle must be between 0 and 1");
		if (sampleRate <= 0)
			throw new IllegalArgumentException("Sample rate must be positive");
		if (duration < 0)
			throw new IllegalArgumentException("Duration must be non-negative");
		if (!mode.equals("BandLimited") && !mode.equals("Naive"))
			throw new IllegalArgumentException("Invalid mode specification");
		
		// Start writing file data
		int numSamples = (int)Math.round(duration * sampleRate);
		try (DataOutputStream out = new DataOutputStream(new BufferedOutputStream(new FileOutputStream(outFile)))) {
			// Write WAV header
			out.write("RIFF".getBytes(StandardCharsets.US_ASCII));
			writeLittleInt32(36 + numSamples * 4, out);
			out.write("WAVE".getBytes(StandardCharsets.US_ASCII));
			out.write("fmt ".getBytes(StandardCharsets.US_ASCII));
			writeLittleInt32(16, out);
			writeLittleInt16(0x0003, out);
			writeLittleInt16(1, out);
			writeLittleInt32(sampleRate, out);
			writeLittleInt32(sampleRate * 4, out);
			writeLittleInt16(4, out);
			writeLittleInt16(32, out);
			out.write("data".getBytes(StandardCharsets.US_ASCII));
			writeLittleInt32(numSamples * 4, out);
			
			if (mode.equals("BandLimited")) {
				// Calculate harmonic amplitudes
				int numHarmonics = (int)(sampleRate / (frequency * 2));
				double[] coefficients = new double[numHarmonics + 1];
				coefficients[0] = dutyCycle - 0.5;  // Start with DC coefficient
				for (int i = 1; i < coefficients.length; i++)
					coefficients[i] = Math.sin(i * dutyCycle * Math.PI) * 2 / (i * Math.PI);
				
				// Generate audio samples
				double scaler = frequency * Math.PI * 2 / sampleRate;
				for (int i = 0; i < numSamples; i++) {
					double temp = scaler * i;
					double val = coefficients[0];
					for (int j = 1; j < coefficients.length; j++)
						val += Math.cos(j * temp) * coefficients[j];
					writeLittleFloat32((float)val, out);
				}
				
			} else if (mode.equals("Naive")) {
				double scaler = frequency / sampleRate;
				double shift = dutyCycle / 2;
				for (int i = 0; i < numSamples; i++) {
					float val = (i * scaler + shift) % 1 < dutyCycle ? 0.5f : -0.5f;
					writeLittleFloat32(val, out);
				}
				
			} else
				throw new AssertionError();
		}
	}
	
	
	
	private static void writeLittleInt16(int val, DataOutput out) throws IOException {
		out.writeShort(Integer.reverseBytes(val) >>> 16);
	}
	
	private static void writeLittleInt32(int val, DataOutput out) throws IOException {
		out.writeInt(Integer.reverseBytes(val));
	}
	
	private static void writeLittleFloat32(float val, DataOutput out) throws IOException {
		out.writeInt(Integer.reverseBytes(Float.floatToRawIntBits(val)));
	}
	
}

File: band-limited-square-waves/generate-square-wave.py
# 
# Band-limited square waves (Python)
# 
# Copyright (c) 2020 Project Nayuki
# All rights reserved. Contact Nayuki for licensing.
# https://www.nayuki.io/page/band-limited-square-waves
# 

import math, pathlib, struct, sys
from typing import List


def main(args: List[str]) -> None:
	# Check number of command line arguments
	if len(args) not in (5, 6):
		sys.exit("""Usage: python generate-square-wave.py Frequency DutyCycle SampleRate Duration [BandLimited/Naive] Output.wav
Example: python generate-square-wave.py 440.0 0.5 48000 1.0 BandLimited Output.wav""")
	
	# Get all command line arguments
	frequency: float = float(args[0])
	dutycycle: float = float(args[1])
	samplerate: int = int(args[2])
	duration: float = float(args[3])
	mode: str
	outfilepath: str
	if len(args) == 5:
		mode = "BandLimited"
		outfilepath = args[4]
	elif len(args) == 6:
		mode = args[4]
		outfilepath = args[5]
	else:
		raise AssertionError()
	
	# Check value ranges
	if frequency <= 0:
		raise ValueError("Frequency must be positive")
	if not (0 <= dutycycle <= 1):
		raise ValueError("Duty cycle must be between 0 and 1")
	if samplerate <= 0:
		raise ValueError("Sample rate must be positive")
	if duration < 0:
		raise ValueError("Duration must be non-negative")
	if mode not in ("BandLimited", "Naive"):
		raise ValueError("Invalid mode specification")
	
	# Start writing file data
	numsamples: int = int(round(duration * samplerate))
	with pathlib.Path(outfilepath).open("wb") as fout:
		# Write WAV header
		fout.write(b"RIFF")
		fout.write(struct.pack("<I", 36 + numsamples * 4))
		fout.write(b"WAVE")
		fout.write(b"fmt ")
		fout.write(struct.pack("<IHHIIHH", 16, 0x0003, 1, samplerate, samplerate * 4, 4, 32))
		fout.write(b"data")
		fout.write(struct.pack("<I", numsamples * 4))
		
		scaler: float
		val: float
		
		if mode == "BandLimited":
			# Calculate harmonic amplitudes
			coefficients: List[float] = [dutycycle - 0.5]  # Start with DC coefficient
			numharmonics: int = int(samplerate // (frequency * 2))
			coefficients += [math.sin(i * dutycycle * math.pi) * 2 / (i * math.pi)
				for i in range(1, numharmonics + 1)]
			
			# Generate audio samples
			scaler = frequency * math.pi * 2 / samplerate
			for i in range(numsamples):
				temp: float = scaler * i
				val = sum(coef * math.cos(j * temp)
					for (j, coef) in enumerate(coefficients))
				fout.write(struct.pack("<f", val))
		
		elif mode == "Naive":
			scaler = frequency / samplerate
			shift: float = dutycycle / 2
			for i in range(numsamples):
				val = 0.5 if ((i * scaler + shift) % 1 < dutycycle) else -0.5
				fout.write(struct.pack("<f", val))
			
		else:
			raise AssertionError()


if __name__ == "__main__":
	main(sys.argv[1 : ])
